src/backend/distributed/executor/multi_client_executor.c

```diff
@@ -315,6 +315,53 @@ MultiClientConnectionUp(int32 connectionId)
 }
 
 
+/* MultiClientExecute synchronously executes a query over the given connection. */
+bool
+MultiClientExecute(int32 connectionId, const char *query, void **queryResult,
+				   int *rowCount, int *columnCount)
+{
+	bool querySent = false;
+	bool queryReady = false;
+	bool queryOK = false;
+	WaitInfo *waitInfo = NULL;
+
+	querySent = MultiClientSendQuery(connectionId, query);
+	if (!querySent)
+	{
+		return false;
+	}
+
+	waitInfo = MultiClientCreateWaitInfo(1);
+
+	while (!queryReady)
+	{
+		ResultStatus resultStatus = MultiClientResultStatus(connectionId);
+		if (resultStatus == CLIENT_RESULT_READY)
+		{
+			queryReady = true;
+		}
+		else if (resultStatus == CLIENT_RESULT_BUSY)
+		{
+			/* wait for results, errors, or interrupts */
+			MultiClientResetWaitInfo(waitInfo);
+			MultiClientRegisterWait(waitInfo, TASK_STATUS_SOCKET_READ, connectionId);
+			MultiClientWait(waitInfo);
```

Comment:
Ah, ok. That's a slightly roundabout way - and I don't think it's entirely reliable (because signals don't reliably interrupt poll(), for that you need something like pg_latch.h's stuff, with the self-pipe trick).  I'm fine with leaving it that way, or with leaving it out...